# <center><span style="color:#64AE2E">Instituto Nacional de Pesquisas Espaciais<br>Programa de Pós Graduação em Sensoriamento Remoto<br>SER-347 - Introdução à Programação para Sensoriamento Remoto</span><center><br><hr style="border:2px solid #DEC827;">

# <span style="color:#F13214">Acurácia do produto "Risco de fogo" no Brasil para diferentes Estados, Biomas, Regiões e Sazonalidade</span>
<hr style="border:2px solid #DE6127;">

<img src="https://uploaddeimagens.com.br/images/002/693/706/original/risco_de_fogo_2019.png?1591449365" alt="Queimadas" style="height: 150px;" align="right">
    
# <span style="color:#DE6127">Pré-processamento</span>

<span style="color:#9C3526">Professores:</span>

<span style="color:#9C3526">- Gilberto Ribeiro de Queiroz</span>

<span style="color:#9C3526">- Thales Sehn Körting</span>

<span style="color:#9C3526">Alunos: </span>

<span style="color:#9C3526">- Bruno Vargas Adorno </span>

<span style="color:#9C3526">- Gabriel Máximo da Silva </span>


Este `Jupyter Notebook` tem a finalidade de detalhar ao usuário as etapas de `preprocessamento` dos dados vetoriais utilizados nesse projeto a fim de 

**i) consolidar as informações de um dado ano que serão necessárias para geração dos resultados em um `Geodataframe`**; 

**ii) gerar `tabelas dinâmicas` preliminares dos dados**.  


As principais etapas de tratamento, destacadas nos markdowns como `riscofogo.py`, foram consolidadas e inseridas em formas de `função` no **módulo** `riscofogo.py`.

A partir do Jupyter notebook `Projeto8_final` podem ser verificadas as etapas de `processamento` dos dados e visualização dos resultados.

> Importando as bibliotecas - `riscofogo.py`

<div class="alert alert-block alert-warning">
⚠️<b> Este projeto utilizou as versões 0.7.0, 1.0.4 e 1.18.4 do geopandas, pandas e numpy, respectivamente.</b>
</div>


In [1]:
import geopandas as gpd # trabalhar com GeoDataFrame 
import pandas as pd # trabalhar com DataFrame
import numpy as np # trabalhar com matrizes homogeneas de multidimensões

In [2]:
print('Versão geopandas {}'.format(gpd.__version__))
print('Versão pandas {}'.format(pd.__version__))
print('Versão numpy {}'.format(np.__version__))

Versão geopandas 0.7.0
Versão pandas 1.0.3
Versão numpy 1.18.1


# i) Consolidação de um Geodataframe

Após abrir o arquivo com a função `read_file`da `geopandas`podemos verificar as informações básicas contidas na tabela

In [3]:
# uso da função read_file da biblioteca geopandas para ler o arquivo shapefile
focos_2015 = gpd.read_file('Dados_focos/2015/Focos_2015-01-01_2015-12-31.shp', encoding='utf-8')


focos_2015



datahora  satelite    pais          estado  \
0       2015/01/01 17:03:00  AQUA_M-T  Brasil           GOIAS   
1       2015/01/01 17:01:00  AQUA_M-T  Brasil  ESPIRITO SANTO   
2       2015/01/01 17:03:00  AQUA_M-T  Brasil    MINAS GERAIS   
3       2015/01/01 17:03:00  AQUA_M-T  Brasil           BAHIA   
4       2015/01/01 17:02:00  AQUA_M-T  Brasil       SAO PAULO   
...                     ...       ...     ...             ...   
216777  2015/12/31 17:31:00  AQUA_M-T  Brasil        AMAZONAS   
216778  2015/12/31 17:31:00  AQUA_M-T  Brasil        AMAZONAS   
216779  2015/12/31 17:31:00  AQUA_M-T  Brasil        AMAZONAS   
216780  2015/12/31 17:31:00  AQUA_M-T  Brasil        AMAZONAS   
216781  2015/12/31 17:31:00  AQUA_M-T  Brasil        AMAZONAS   

               municipio           bioma diasemchuv  precipitac  riscofogo  \
0                JUSSARA         Cerrado       None         0.0       0.19   
1       DOMINGOS MARTINS  Mata Atlantica       None         0.0       1.00   
2               PARACATU         Cerrado       None         0.0       0.90   
3                  COCOS         Cerrado       None         0.0       0.03   
4                 PONTAL         Cerrado       None         0.0       0.35   
...                  ...             ...        ...         ...        ...   
216777       BARREIRINHA        Amazonia       None         0.0        NaN   
216778       BARREIRINHA        Amazonia       None         0.0        NaN   
216779         PARINTINS        Amazonia       None         0.0        NaN   
216780       BARREIRINHA        Amazonia       None         0.0        NaN   
216781             MAUES        Amazonia       None        11.1        NaN   

        latitude  longitude   frp                     geometry  
0        -15.302    -51.558  None  POINT (-51.55800 -15.30200)  
1        -20.357    -40.938  None  POINT (-40.93800 -20.35700)  
2        -17.170    -46.415  None  POINT (-46.41500 -17.17000)  
3        -15.086    -45.999  None  POINT (-45.99900 -15.08600)  
4        -21.061    -48.023  None  POINT (-48.02300 -21.06100)  
...          ...        ...   ...                          ...  
216777    -2.908    -57.066  None   POINT (-57.06600 -2.90800)  
216778    -2.906    -57.057  None   POINT (-57.05700 -2.90600)  
216779    -2.549    -57.325  None   POINT (-57.32500 -2.54900)  
216780    -2.678    -57.395  None   POINT (-57.39500 -2.67800)  
216781    -3.458    -57.564  None   POINT (-57.56400 -3.45800)  

[216782 rows x 13 columns]

Verificando a projeção do arquivo `focos_2015` -> `EPSG:4326`

In [4]:
focos_2015.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

Pode-se observar que há colunas com dados vazios em todas as 216782 linhas (`diasemchuv`e `frp`), enquanto a coluna `riscofogo`possui células sem dados em `4,3%` das linhas. 

Como a ideia é trabalhar com amostragem aleatória, essa perda de dados não foi considerada um problema, a princípio, em meio a grande quantidade de observações.

In [5]:
# observando ocorrência de valores vazios NaN
focos_2015.isnull().sum()

datahora           0
satelite           0
pais               0
estado             0
municipio          0
bioma              0
diasemchuv    216782
precipitac         0
riscofogo       9430
latitude           0
longitude          0
frp           216782
geometry           0
dtype: int64

Aqui, para evitar conflitos de dados, foram eliminadas as linhas da coluna `riscofogo`sem dados, bem como as colunas `diasemchuv`e `frp`que também continham células vazias. Adicionalmente, pela função `sample`uma amostra aleatória de `10%` dos dados restantes (207.352 linhas) foi realizada.

> - A função 'dropna' elimina as linhas com valores sem dados (NaN ou None) de uma coluna indicada no argumento subset.
> - A função 'drop' exclui colunas indicadas em uma lista, inserida no argumento 'columns'
> - A função 'sample' permite amostrar os dados aleatoriamente, com pesos que variam de 0-1, sendo que mais próximo de 1, maior número de dados amostrados. - **`riscofogo.py`**

In [6]:
focos_2015_final =  focos_2015.dropna(subset=['riscofogo']).drop(columns=['frp','diasemchuv']).sample(frac=0.1, replace=True, random_state=1)


focos_2015_final

datahora  satelite    pais        estado  \
132365  2015/09/21 17:11:00  AQUA_M-T  Brasil   MATO GROSSO   
5494    2015/07/13 17:47:00  AQUA_M-T  Brasil   MATO GROSSO   
52694   2015/10/27 16:45:00  AQUA_M-T  Brasil         BAHIA   
113554  2015/09/08 16:01:00  AQUA_M-T  Brasil  MINAS GERAIS   
77576   2015/07/12 17:05:00  AQUA_M-T  Brasil         PIAUI   
...                     ...       ...     ...           ...   
2752    2015/03/14 17:56:00  AQUA_M-T  Brasil      AMAZONAS   
92166   2015/08/15 16:53:00  AQUA_M-T  Brasil          PARA   
50123   2015/10/21 17:22:00  AQUA_M-T  Brasil   MATO GROSSO   
97664   2015/08/20 18:50:00  AQUA_M-T  Brasil          ACRE   
179473  2015/11/09 17:54:00  AQUA_M-T  Brasil      RONDONIA   

                      municipio     bioma  precipitac  riscofogo  latitude  \
132365                  COLNIZA  Amazonia         0.0       1.00    -9.265   
5494               NOVA UBIRATA   Cerrado         0.0       1.00   -12.875   
52694    TABOCAS DO BREJO VELHO   Cerrado         0.0       1.00   -12.476   
113554                 JANUARIA   Cerrado         0.0       1.00   -15.091   
77576   BAIXA GRANDE DO RIBEIRO   Cerrado         0.0       1.00    -8.892   
...                         ...       ...         ...        ...       ...   
2752                       TEFE  Amazonia         1.8       0.01    -3.357   
92166        SAO FELIX DO XINGU  Amazonia         0.0       1.00    -6.162   
50123              NOVA MARINGA  Amazonia         0.0       0.51   -12.915   
97664                  TARAUACA  Amazonia         0.3       0.38    -7.974   
179473              PORTO VELHO  Amazonia         0.2       1.00    -9.094   

        longitude                     geometry  
132365    -60.932   POINT (-60.93200 -9.26500)  
5494      -54.895  POINT (-54.89500 -12.87500)  
52694     -44.216  POINT (-44.21600 -12.47600)  
113554    -45.068  POINT (-45.06800 -15.09100)  
77576     -45.272   POINT (-45.27200 -8.89200)  
...           ...                          ...  
2752      -64.720   POINT (-64.72000 -3.35700)  
92166     -52.086   POINT (-52.08600 -6.16200)  
50123     -56.912  POINT (-56.91200 -12.91500)  
97664     -70.960   POINT (-70.96000 -7.97400)  
179473    -64.121   POINT (-64.12100 -9.09400)  

[20735 rows x 11 columns]

Pode ser confirmada ausência de dados nulos

In [7]:
print('Dados nulos\n')
print(focos_2015_final.isnull().sum())

Dados nulos

datahora      0
satelite      0
pais          0
estado        0
municipio     0
bioma         0
precipitac    0
riscofogo     0
latitude      0
longitude     0
geometry      0
dtype: int64


Aqui pode ser verificado os tipos de dados das colunas do `Geodataframe`. 

In [8]:
print('Tipos dos dados\n')
print(focos_2015_final.dtypes)

Tipos dos dados

datahora        object
satelite        object
pais            object
estado          object
municipio       object
bioma           object
precipitac     float64
riscofogo      float64
latitude       float64
longitude      float64
geometry      geometry
dtype: object


> Para iterar e fazer operações com a variação temporal no `Dataframe` ou `Geodataframe`, a função `to_datetime` do `Pandas` permite alterar o tipo de dado `objeto` para `'datetime64` - **`riscofogo.py`**


In [9]:
focos_2015_final['datahora'] = pd.to_datetime(focos_2015_final['datahora']) 
focos_2015_final.dtypes

datahora      datetime64[ns]
satelite              object
pais                  object
estado                object
municipio             object
bioma                 object
precipitac           float64
riscofogo            float64
latitude             float64
longitude            float64
geometry            geometry
dtype: object

Criando uma coluna de `dia do ano` e então de `estações`.
> Optou-se por criar uma coluna de `dia do ano`, por meio da função `dt.dayofyear` do `Pandas` aplicada sobre uma `série` `datahora`  para facilitar a definição das estações correspondentes pela mesma função `select` do `numpy` - `riscofogo.py`

In [10]:
focos_2015_final['dia do ano'] = focos_2015_final['datahora'].dt.dayofyear



conditions_estacao = [
    focos_2015_final['dia do ano'] > 356, 
    focos_2015_final['dia do ano'] > 264,
    focos_2015_final['dia do ano'] > 172,
    focos_2015_final['dia do ano'] > 80]
choices_estacao = ['VERÃO','PRIMAVERA','INVERNO','OUTONO']
focos_2015_final['estacoes'] = np.select(conditions_estacao, choices_estacao, default='VERÃO')


focos_2015_final

datahora  satelite    pais        estado  \
132365 2015-09-21 17:11:00  AQUA_M-T  Brasil   MATO GROSSO   
5494   2015-07-13 17:47:00  AQUA_M-T  Brasil   MATO GROSSO   
52694  2015-10-27 16:45:00  AQUA_M-T  Brasil         BAHIA   
113554 2015-09-08 16:01:00  AQUA_M-T  Brasil  MINAS GERAIS   
77576  2015-07-12 17:05:00  AQUA_M-T  Brasil         PIAUI   
...                    ...       ...     ...           ...   
2752   2015-03-14 17:56:00  AQUA_M-T  Brasil      AMAZONAS   
92166  2015-08-15 16:53:00  AQUA_M-T  Brasil          PARA   
50123  2015-10-21 17:22:00  AQUA_M-T  Brasil   MATO GROSSO   
97664  2015-08-20 18:50:00  AQUA_M-T  Brasil          ACRE   
179473 2015-11-09 17:54:00  AQUA_M-T  Brasil      RONDONIA   

                      municipio     bioma  precipitac  riscofogo  latitude  \
132365                  COLNIZA  Amazonia         0.0       1.00    -9.265   
5494               NOVA UBIRATA   Cerrado         0.0       1.00   -12.875   
52694    TABOCAS DO BREJO VELHO   Cerrado         0.0       1.00   -12.476   
113554                 JANUARIA   Cerrado         0.0       1.00   -15.091   
77576   BAIXA GRANDE DO RIBEIRO   Cerrado         0.0       1.00    -8.892   
...                         ...       ...         ...        ...       ...   
2752                       TEFE  Amazonia         1.8       0.01    -3.357   
92166        SAO FELIX DO XINGU  Amazonia         0.0       1.00    -6.162   
50123              NOVA MARINGA  Amazonia         0.0       0.51   -12.915   
97664                  TARAUACA  Amazonia         0.3       0.38    -7.974   
179473              PORTO VELHO  Amazonia         0.2       1.00    -9.094   

        longitude                     geometry  dia do ano   estacoes  
132365    -60.932   POINT (-60.93200 -9.26500)         264    INVERNO  
5494      -54.895  POINT (-54.89500 -12.87500)         194    INVERNO  
52694     -44.216  POINT (-44.21600 -12.47600)         300  PRIMAVERA  
113554    -45.068  POINT (-45.06800 -15.09100)         251    INVERNO  
77576     -45.272   POINT (-45.27200 -8.89200)         193    INVERNO  
...           ...                          ...         ...        ...  
2752      -64.720   POINT (-64.72000 -3.35700)          73      VERÃO  
92166     -52.086   POINT (-52.08600 -6.16200)         227    INVERNO  
50123     -56.912  POINT (-56.91200 -12.91500)         294  PRIMAVERA  
97664     -70.960   POINT (-70.96000 -7.97400)         232    INVERNO  
179473    -64.121   POINT (-64.12100 -9.09400)         313  PRIMAVERA  

[20735 rows x 13 columns]

Verificando a distribuição total de focos por estação

In [11]:
focos_2015_estacao = focos_2015_final.groupby('estacoes').estacoes.count()

focos_2015_estacao

estacoes
INVERNO      9064
OUTONO        836
PRIMAVERA    9838
VERÃO         997
Name: estacoes, dtype: int64

Criando uma coluna com níveis de riscofogo, de acordo com definido pelos autores _Setzer et al.(2019)_
> A partir de uma lista de condições e de escolhas a função `select` da numpy permite selecionar `condições` de uma série e determinar uma `escolha de valor` correspondente em uma série nova; o argumento `default` estabelece a escolha para os valores que não tiverem condições definidas. - `riscofogo.py`

In [12]:
import numpy as np

conditions = [
    focos_2015_final['riscofogo'] > 0.95,
    focos_2015_final['riscofogo'] > 0.7,
    focos_2015_final['riscofogo'] > 0.4,
    focos_2015_final['riscofogo'] > 0.15]
choices = ['CRITICO', 'ALTO', 'MEDIO', 'BAIXO']
focos_2015_final['risco_classe'] = np.select(conditions, choices, default='MINIMO')

# criando coluna de legenda para o mapa
conditions_legenda = [
    focos_2015_final['riscofogo'] > 0.95,
    focos_2015_final['riscofogo'] > 0.7,
    focos_2015_final['riscofogo'] > 0.4,
    focos_2015_final['riscofogo'] > 0.15]
choices_legenda = ['1','2', '3', '4']
focos_2015_final['legenda'] = np.select(conditions_legenda, choices_legenda, default=5)
    

focos_2015_final

datahora  satelite    pais        estado  \
132365 2015-09-21 17:11:00  AQUA_M-T  Brasil   MATO GROSSO   
5494   2015-07-13 17:47:00  AQUA_M-T  Brasil   MATO GROSSO   
52694  2015-10-27 16:45:00  AQUA_M-T  Brasil         BAHIA   
113554 2015-09-08 16:01:00  AQUA_M-T  Brasil  MINAS GERAIS   
77576  2015-07-12 17:05:00  AQUA_M-T  Brasil         PIAUI   
...                    ...       ...     ...           ...   
2752   2015-03-14 17:56:00  AQUA_M-T  Brasil      AMAZONAS   
92166  2015-08-15 16:53:00  AQUA_M-T  Brasil          PARA   
50123  2015-10-21 17:22:00  AQUA_M-T  Brasil   MATO GROSSO   
97664  2015-08-20 18:50:00  AQUA_M-T  Brasil          ACRE   
179473 2015-11-09 17:54:00  AQUA_M-T  Brasil      RONDONIA   

                      municipio     bioma  precipitac  riscofogo  latitude  \
132365                  COLNIZA  Amazonia         0.0       1.00    -9.265   
5494               NOVA UBIRATA   Cerrado         0.0       1.00   -12.875   
52694    TABOCAS DO BREJO VELHO   Cerrado         0.0       1.00   -12.476   
113554                 JANUARIA   Cerrado         0.0       1.00   -15.091   
77576   BAIXA GRANDE DO RIBEIRO   Cerrado         0.0       1.00    -8.892   
...                         ...       ...         ...        ...       ...   
2752                       TEFE  Amazonia         1.8       0.01    -3.357   
92166        SAO FELIX DO XINGU  Amazonia         0.0       1.00    -6.162   
50123              NOVA MARINGA  Amazonia         0.0       0.51   -12.915   
97664                  TARAUACA  Amazonia         0.3       0.38    -7.974   
179473              PORTO VELHO  Amazonia         0.2       1.00    -9.094   

        longitude                     geometry  dia do ano   estacoes  \
132365    -60.932   POINT (-60.93200 -9.26500)         264    INVERNO   
5494      -54.895  POINT (-54.89500 -12.87500)         194    INVERNO   
52694     -44.216  POINT (-44.21600 -12.47600)         300  PRIMAVERA   
113554    -45.068  POINT (-45.06800 -15.09100)         251    INVERNO   
77576     -45.272   POINT (-45.27200 -8.89200)         193    INVERNO   
...           ...                          ...         ...        ...   
2752      -64.720   POINT (-64.72000 -3.35700)          73      VERÃO   
92166     -52.086   POINT (-52.08600 -6.16200)         227    INVERNO   
50123     -56.912  POINT (-56.91200 -12.91500)         294  PRIMAVERA   
97664     -70.960   POINT (-70.96000 -7.97400)         232    INVERNO   
179473    -64.121   POINT (-64.12100 -9.09400)         313  PRIMAVERA   

       risco_classe legenda  
132365      CRITICO       1  
5494        CRITICO       1  
52694       CRITICO       1  
113554      CRITICO       1  
77576       CRITICO       1  
...             ...     ...  
2752         MINIMO       5  
92166       CRITICO       1  
50123         MEDIO       3  
97664         BAIXO       4  
179473      CRITICO       1  

[20735 rows x 15 columns]

> Abrindo arquivo shape de unidades federativas para incluir dados de regiao no dataframe - `riscofogo.py`

In [13]:
UF = gpd.read_file('uf-2018/BRUFE250GC_SIR.shp', encoding='utf-8')

UF

NM_ESTADO     NM_REGIAO CD_GEOCUF  \
0               SERGIPE      NORDESTE        28   
1              MARANHÃO      NORDESTE        21   
2        ESPÍRITO SANTO       SUDESTE        32   
3              AMAZONAS         NORTE        13   
4               RORAIMA         NORTE        14   
5                 GOIÁS  CENTRO-OESTE        52   
6                 AMAPÁ         NORTE        16   
7     RIO GRANDE DO SUL           SUL        43   
8               PARAÍBA      NORDESTE        25   
9                 PIAUÍ      NORDESTE        22   
10            SÃO PAULO       SUDESTE        35   
11       SANTA CATARINA           SUL        42   
12           PERNAMBUCO      NORDESTE        26   
13       RIO DE JANEIRO       SUDESTE        33   
14   MATO GROSSO DO SUL  CENTRO-OESTE        50   
15          MATO GROSSO  CENTRO-OESTE        51   
16                BAHIA      NORDESTE        29   
17         MINAS GERAIS       SUDESTE        31   
18              ALAGOAS      NORDESTE        27   
19                CEARÁ      NORDESTE        23   
20  RIO GRANDE DO NORTE      NORDESTE        24   
21               PARANÁ           SUL        41   
22             RONDÔNIA         NORTE        11   
23     DISTRITO FEDERAL  CENTRO-OESTE        53   
24                 ACRE         NORTE        12   
25                 PARÁ         NORTE        15   
26            TOCANTINS         NORTE        17   

                                             geometry  
0   MULTIPOLYGON (((-37.98248 -11.34818, -37.98253...  
1   MULTIPOLYGON (((-44.48155 -2.72660, -44.47867 ...  
2   MULTIPOLYGON (((-29.33711 -20.48802, -29.33686...  
3   POLYGON ((-69.61341 -2.44025, -69.53446 -2.000...  
4   POLYGON ((-63.97805 2.46754, -63.97871 2.46840...  
5   POLYGON ((-49.36953 -13.27462, -49.35361 -13.2...  
6   MULTIPOLYGON (((-53.27918 2.18599, -53.27655 2...  
7   MULTIPOLYGON (((-49.70392 -29.34774, -49.70402...  
8   MULTIPOLYGON (((-34.79576 -7.35000, -34.79601 ...  
9   POLYGON ((-42.91539 -6.64607, -42.91645 -6.644...  
10  MULTIPOLYGON (((-48.03781 -25.35643, -48.03773...  
11  MULTIPOLYGON (((-48.34793 -27.44124, -48.34767...  
12  MULTIPOLYGON (((-32.39125 -3.81747, -32.39110 ...  
13  MULTIPOLYGON (((-41.78720 -22.76892, -41.78684...  
14  POLYGON ((-57.51683 -22.17363, -57.51741 -22.1...  
15  POLYGON ((-60.38233 -13.99242, -60.38206 -13.9...  
16  MULTIPOLYGON (((-38.70118 -17.97917, -38.70125...  
17  POLYGON ((-47.28285 -18.04089, -47.28330 -18.0...  
18  MULTIPOLYGON (((-37.13342 -9.91321, -37.14254 ...  
19  MULTIPOLYGON (((-38.79615 -3.53504, -38.79830 ...  
20  MULTIPOLYGON (((-37.04412 -4.81854, -37.04501 ...  
21  MULTIPOLYGON (((-48.36381 -25.73616, -48.36408...  
22  POLYGON ((-62.89216 -12.85884, -62.89371 -12.8...  
23  POLYGON ((-47.81455 -16.04995, -47.81478 -16.0...  
24  POLYGON ((-71.07772 -9.82774, -71.07817 -9.828...  
25  MULTIPOLYGON (((-46.31362 -0.95914, -46.31382 ...  
26  POLYGON ((-48.14665 -13.15201, -48.14796 -13.1...

Verificando a projeção do arquivo `UF` -> `EPSG: 4674`, diferente da projeção do arquivo `focos de calor`

In [14]:
UF.crs

<Geographic 2D CRS: EPSG:4674>
Name: SIRGAS 2000
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: Latin America - SIRGAS 2000 by country
- bounds: (-122.19, -59.87, -25.28, 32.72)
Datum: Sistema de Referencia Geocentrico para las AmericaS 2000
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

> Como já tínhamos coluna de `estado`, e `CD_GEOCUF` não seria importante, as mesmas foram eliminadas. - `riscofogo.py`

In [15]:
UF_Regiao = UF.drop(columns=['NM_ESTADO','CD_GEOCUF'])
UF_Regiao

NM_REGIAO                                           geometry
0       NORDESTE  MULTIPOLYGON (((-37.98248 -11.34818, -37.98253...
1       NORDESTE  MULTIPOLYGON (((-44.48155 -2.72660, -44.47867 ...
2        SUDESTE  MULTIPOLYGON (((-29.33711 -20.48802, -29.33686...
3          NORTE  POLYGON ((-69.61341 -2.44025, -69.53446 -2.000...
4          NORTE  POLYGON ((-63.97805 2.46754, -63.97871 2.46840...
5   CENTRO-OESTE  POLYGON ((-49.36953 -13.27462, -49.35361 -13.2...
6          NORTE  MULTIPOLYGON (((-53.27918 2.18599, -53.27655 2...
7            SUL  MULTIPOLYGON (((-49.70392 -29.34774, -49.70402...
8       NORDESTE  MULTIPOLYGON (((-34.79576 -7.35000, -34.79601 ...
9       NORDESTE  POLYGON ((-42.91539 -6.64607, -42.91645 -6.644...
10       SUDESTE  MULTIPOLYGON (((-48.03781 -25.35643, -48.03773...
11           SUL  MULTIPOLYGON (((-48.34793 -27.44124, -48.34767...
12      NORDESTE  MULTIPOLYGON (((-32.39125 -3.81747, -32.39110 ...
13       SUDESTE  MULTIPOLYGON (((-41.78720 -22.76892, -41.78684...
14  CENTRO-OESTE  POLYGON ((-57.51683 -22.17363, -57.51741 -22.1...
15  CENTRO-OESTE  POLYGON ((-60.38233 -13.99242, -60.38206 -13.9...
16      NORDESTE  MULTIPOLYGON (((-38.70118 -17.97917, -38.70125...
17       SUDESTE  POLYGON ((-47.28285 -18.04089, -47.28330 -18.0...
18      NORDESTE  MULTIPOLYGON (((-37.13342 -9.91321, -37.14254 ...
19      NORDESTE  MULTIPOLYGON (((-38.79615 -3.53504, -38.79830 ...
20      NORDESTE  MULTIPOLYGON (((-37.04412 -4.81854, -37.04501 ...
21           SUL  MULTIPOLYGON (((-48.36381 -25.73616, -48.36408...
22         NORTE  POLYGON ((-62.89216 -12.85884, -62.89371 -12.8...
23  CENTRO-OESTE  POLYGON ((-47.81455 -16.04995, -47.81478 -16.0...
24         NORTE  POLYGON ((-71.07772 -9.82774, -71.07817 -9.828...
25         NORTE  MULTIPOLYGON (((-46.31362 -0.95914, -46.31382 ...
26         NORTE  POLYGON ((-48.14665 -13.15201, -48.14796 -13.1...

> Para garantir exata correspondência entre as geometrias dos shapes `UF_Regiao`e `focos_2015_final`, o primeiro foi reprojetado para `EPSG:4326` - `riscofogo.py`

In [16]:
UF_Regiao_WGS84 = UF_Regiao.to_crs('EPSG:4326')

#reverificando crs

UF_Regiao_WGS84.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

> A função do GeoPandas `sjoin` com a opção `intersect` transporta as séries de dados do `geodataframe` no segundo argumento para  o primeiro a partir da análise de suas `geometrias`. - `riscofogo.py`

In [17]:
#inserindo info das regiões no shape de focos.
focos_2015_final = gpd.sjoin(focos_2015_final, UF_Regiao_WGS84, how='inner', op='intersects')

focos_2015_final


datahora  satelite    pais            estado  \
132365 2015-09-21 17:11:00  AQUA_M-T  Brasil       MATO GROSSO   
5494   2015-07-13 17:47:00  AQUA_M-T  Brasil       MATO GROSSO   
22272  2015-09-08 17:40:00  AQUA_M-T  Brasil       MATO GROSSO   
180910 2015-11-11 17:40:00  AQUA_M-T  Brasil       MATO GROSSO   
180006 2015-11-10 16:57:00  AQUA_M-T  Brasil       MATO GROSSO   
...                    ...       ...     ...               ...   
16880  2015-08-22 16:56:00  AQUA_M-T  Brasil  DISTRITO FEDERAL   
79197  2015-07-14 16:51:00  AQUA_M-T  Brasil  DISTRITO FEDERAL   
154734 2015-10-11 16:44:00  AQUA_M-T  Brasil  DISTRITO FEDERAL   
17650  2015-08-29 17:03:00  AQUA_M-T  Brasil  DISTRITO FEDERAL   
209204 2015-09-18 16:38:00  AQUA_M-T  Brasil  DISTRITO FEDERAL   

                               municipio     bioma  precipitac  riscofogo  \
132365                           COLNIZA  Amazonia         0.0       1.00   
5494                        NOVA UBIRATA   Cerrado         0.0       1.00   
22272                           CANARANA   Cerrado         0.0       0.54   
180910  VILA BELA DA SANTISSIMA TRINDADE  Amazonia         0.0       0.59   
180006              RIBEIRAO CASCALHEIRA   Cerrado         0.1       0.00   
...                                  ...       ...         ...        ...   
16880                           BRASILIA   Cerrado         0.0       1.00   
79197                           BRASILIA   Cerrado         0.0       1.00   
154734                          BRASILIA   Cerrado         0.0       1.00   
17650                           BRASILIA   Cerrado         0.1       1.00   
209204                          BRASILIA   Cerrado         0.0       1.00   

        latitude  longitude                     geometry  dia do ano  \
132365    -9.265    -60.932   POINT (-60.93200 -9.26500)         264   
5494     -12.875    -54.895  POINT (-54.89500 -12.87500)         194   
22272    -13.428    -51.803  POINT (-51.80300 -13.42800)         251   
180910   -15.396    -59.832  POINT (-59.83200 -15.39600)         315   
180006   -12.727    -51.239  POINT (-51.23900 -12.72700)         314   
...          ...        ...                          ...         ...   
16880    -15.930    -47.592  POINT (-47.59200 -15.93000)         234   
79197    -15.850    -48.269  POINT (-48.26900 -15.85000)         195   
154734   -15.752    -47.651  POINT (-47.65100 -15.75200)         284   
17650    -15.973    -47.994  POINT (-47.99400 -15.97300)         241   
209204   -15.610    -48.124  POINT (-48.12400 -15.61000)         261   

         estacoes risco_classe legenda  index_right     NM_REGIAO  
132365    INVERNO      CRITICO       1           15  CENTRO-OESTE  
5494      INVERNO      CRITICO       1           15  CENTRO-OESTE  
22272     INVERNO        MEDIO       3           15  CENTRO-OESTE  
180910  PRIMAVERA        MEDIO       3           15  CENTRO-OESTE  
180006  PRIMAVERA       MINIMO       5           15  CENTRO-OESTE  
...           ...          ...     ...          ...           ...  
16880     INVERNO      CRITICO       1           23  CENTRO-OESTE  
79197     INVERNO      CRITICO       1           23  CENTRO-OESTE  
154734  PRIMAVERA      CRITICO       1           23  CENTRO-OESTE  
17650     INVERNO      CRITICO       1           23  CENTRO-OESTE  
209204    INVERNO      CRITICO       1           23  CENTRO-OESTE  

[20735 rows x 17 columns]

# ii) Tabelas dinâmicas

A  próxima etapa é a criarão de uma `tabela dinâmica` a partir da função do `Pandas` `pivot_table`, para cada delimitação geográfica que se pretende analisar o acerto do risco de fogo

Como argumento da função `pivot_table`, temos na sequência: 

1- o (Geo)dataframe de entrada; 

2- a coluna cujos valores servirão de base para a operação definida no argumento `aggfunc`; 

3- a coluna do (Geo)dataframe que vai compor os índices (as linhas) da tabela dinâmica; 

4- a coluna do geodataframe que vai compor as colunas da tabela dinâmica; 

5- a operação que será realizada sobre os valores definidos em 2 (que no caso é a contagem - atribui-se valor 1 
para cada célula com dados);

6- valor para preencher células com valores que faltam; 

7 - opção se quer apresentar o somatório das linhas e colunas na tabela dinâmica.

- `riscofogo.py`

In [18]:
pvt_estacoes = pd.pivot_table(focos_2015_final, values=['riscofogo'], index=['estacoes'],
                     columns=['risco_classe'], aggfunc=['count'],
                     fill_value=0, margins=False)

pvt_estacoes

count                           
             riscofogo                           
risco_classe      ALTO BAIXO CRITICO MEDIO MINIMO
estacoes                                         
INVERNO            995   518    6140   888    523
OUTONO              99   131     302   138    166
PRIMAVERA         1044   777    6018   850   1149
VERÃO              162   143     280   108    304

Para as etapas de processamento e visualização dos resultados, direcione-se para o `Jupyter Notebook`:`Projeto8_final`